### PR TITLE
DataSourceSrv: Look up data source by uid and name transparently

### PR DESCRIPTION
**Pull request by torkelo**
This is related to work to make query editors and query execution a little more future-proof and less stuck in how things have been for ever in dashboards where data source always was referred to by name in the persisted models. 

This should make it easier to start using data source uid as the primary way to lookup data sources. The biggest problem with the current approach is that if you change a data source name dashboards and panels do not update (and become broken).  This PR does not fix that but lays some ground work (to keep scope smaller) by making DataSourceSrv a bit more robust (with more unit tests) and support for uid & name looks in the same get function. Thinking is we first check if name matches a uid then assume you're fetching a data source by uid. This way we could potentially start changing panel.datasource to be the uid while being backward compatible with name.

Question is when we change from data source name to uid, it will make the json harder to reason about for users who use dashboard json & provisioning. They could continue using name but after a change in data source picker in Grafana it would switch it to uid. Think that will be acceptable, but can be discussed when we do that change. 

This PR also tries to clean up some access to config.datasources (think none should use this and long term this should go away) but some remain to be removed in follow-up PRs

Related #1664

### CODE REVIEW

File: packages/grafana-runtime/src/utils/DataSourceWithBackend.ts [hunk 1]
```diff
@@ -57,34 +57,37 @@ export class DataSourceWithBackend<
...
     }
+
     const queries = targets.map(q => {
       let datasourceId = this.id;
+
       if (q.datasource === ExpressionDatasourceID) {
         return {
           ...q,
           datasourceId,
-          orgId,
```
Comment (torkelo) [inline packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:71]: @ryantxu I did not get this, why is orgId part of each query? it was not used in https://github.com/grafana/grafana/blob/master/pkg/api/metrics.go#L61. And if used it needs to be added in the backend, the frontend value cannot be trusted, or is it just for tracking/logging?

Comment (ryantxu) [reply packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:71]: Looks like it can be safely removed.

**Review by ryantxu — APPROVED** [review]
Looks good -- can/should we also deprecate:

```
  getDataSourceSettingsByUid(uid: string): DataSourceInstanceSettings | undefined;
```

and recommend `getSettingsFor`?  The function seems to mostly be used in tests

File: packages/grafana-runtime/src/services/dataSourceSrv.ts [hunk 2]
```diff
@@ -29,6 +29,11 @@ export interface DataSourceSrv {
    * Get all data sources except for internal ones that usually should not be listed like mixed data source.
    */
   getExternal(): DataSourceInstanceSettings[];
+
+  /**
+   * Get settings by name or uid
+   */
+  getSettingsFor(nameOrUid: string | null | undefined): DataSourceInstanceSettings | undefined;
```
Comment (hugohaggmark) [inline packages/grafana-runtime/src/services/dataSourceSrv.ts:36]: I'm a bit confused about the api for `DataSourceSrv `:

Should we deprecate getDataSourceSettingsByUid or have 2 distinct functions? One for uid and one for name? I understand that name is going away, but I think having both uid and name in the same argument is confusing for consumers.

Comment (torkelo) [reply packages/grafana-runtime/src/services/dataSourceSrv.ts:36]: I have another PR that touches some stuff related to this function (for a feature it was created for). Was hoping to remove it but might have to deprecate it for a version first
https://github.com/grafana/grafana/pull/29447

**Pull request merged**